src/extension/geofence/src/main/java/org/geoserver/geofence/GeofenceAccessManager.java

```diff
@@ -269,108 +279,189 @@ public StyleAccessLimits getAccessLimits(Authentication user, StyleInfo style) {
+            allowedArea =
+                    getReprojectArea(accessInfo, lgi.getBounds().getCoordinateReferenceSystem());
         }
+        return new ImmutablePair<>(allowedArea, lessRestrictiveRule);
+    }
 
-        DataAccessLimits limits = buildAccessLimits(resource, rule);
-        LOGGER.log(
-                Level.FINE,
-                "Returning {0} for layer {1} and user {2}",
-                new Object[] {limits, resource.prefixedName(), username});
+    // get the AccessInfo from the layer's containers returning a tuple with
+    // the Geometry of the allowed area if found and the less restrictive accessInfo
+    // among the ones of associated layerGroups
+    private Pair<Geometry, AccessInfo> getContainerRuleForLayerDirectAccess(
+            Object resource, Authentication user) {
+        MutablePair<Geometry, AccessInfo> result = new MutablePair<>(null, null);
+        Collection<LayerGroupContainmentCache.LayerGroupSummary> summaries;
+        if (resource instanceof ResourceInfo)
+            summaries = groupsCache.getContainerGroupsFor((ResourceInfo) resource);
+        else if (resource instanceof LayerInfo)
+            summaries = groupsCache.getContainerGroupsFor(((LayerInfo) resource).getResource());
+        else summaries = groupsCache.getContainerGroupsFor((LayerGroupInfo) resource);
+
+        for (LayerGroupContainmentCache.LayerGroupSummary gs : summaries) {
+            LayerGroupInfo.Mode mode = gs.getMode();
+            if (mode.equals(LayerGroupInfo.Mode.OPAQUE_CONTAINER)) {
+                // opaque mode deny access for the layer
+                AccessInfo newInfo = AccessInfo.DENY_ALL;
+                result.setValue(getLessRestrictiveRule(result.getRight(), newInfo));
+            } else if (!mode.equals(LayerGroupInfo.Mode.SINGLE)) {
+                // not opaque and not single mode, the container rule
+                // should override the layer rule
+                String workspace = gs.getWorkspace();
+                String layer = gs.getName();
+                RuleFilter filter = buildRuleFilter(workspace, layer, user);
+                AccessInfo newInfo = rules.getAccessInfo(filter);
+                newInfo = getLessRestrictiveRule(result.getRight(), newInfo);
+                result.setRight(newInfo);
+                if (newInfo.getAreaWkt() != null) {
+                    LayerGroupInfo gi = catalog.getLayerGroupByName(layer);
+                    CoordinateReferenceSystem crs = gi.getBounds().getCoordinateReferenceSystem();
+                    Geometry allowedArea = getReprojectArea(newInfo, crs);
+                    result.setLeft(allowedArea);
+                }
+            }
+        }
+        return result;
+    }
 
-        return limits;
+    // compares two access info and return the less restrictive one
+    private AccessInfo getLessRestrictiveRule(AccessInfo current, AccessInfo newInfo) {
+        if (newInfo == null) return current;
```

Comment:
Since each branch contains a return stratement, no need for all the "else"

Reply:
fixed